### ISSUE TITLE
FileNotFoundError: [Errno 2] No such file or directory: '-'

### ISSUE THREAD

**Issue by jmakitalo** [pyinstaller/pyinstaller#4034]
This error occurs when bundling application with `noarchive=True` . Without this option the bundling is successful and the application works.

The problem seems to occur if there are namespaces in a python package i.e. some subdirectories don't have `__init__.py`. I added such namespaces to a minimal example to reproduce the issue. The real problem is that some standard Python packages also use namespaces in my actual project and these prevent using Pyinstaller with `noarchive=True`. I need this option to be able to use special HASP encryption applied to individual pyc files.

I'm quite sure that this has to do with using `-` as package identifier. This has been done in `writers.py` line 193: `if path in ('-', None):`. However, in `utils/misc.py` in function `compile_py_files` there is no such if-clause even  though `fnm` turns out to be `-` at some point.

# Environment

- Python 3.5.1 32-bit 
- Windows 10
- In virtual environment set up with `py -3.5-32 -m venv venv`
- Installed Pyinstaller 3.4 by `pip install pyinstaller`
- Tried also the develop version of pyinstaller and the problem remains

# Minimal example

Attached as zip file as this has to do with file structure.

[pyinstaller_bug.zip](https://github.com/pyinstaller/pyinstaller/files/2821106/pyinstaller_bug.zip)

To run:
```
py -3.5-32 -m venv venv
venv\Scripts\activate
pip install pyinstaller==3.4
pyinstaller main.spec

```

# Output

```
(venv) D:\pyinstaller_bug>pyinstaller main.spec
72 INFO: PyInstaller: 3.5.dev0+baf94fc1b
72 INFO: Python: 3.5.1
72 INFO: Platform: Windows-10-10.0.16299-SP0
76 INFO: UPX is not available.
78 INFO: Extending PYTHONPATH with paths
['D:\\pyinstaller_bug', 'D:\\pyinstaller_bug']
78 INFO: checking Analysis
78 INFO: Building Analysis because Analysis-00.toc is non existent
78 INFO: Initializing module dependency graph...
82 INFO: Initializing module graph hooks...
84 INFO: Analyzing base_library.zip ...
3830 INFO: running Analysis Analysis-00.toc
4060 WARNING: lib not found: api-ms-win-crt-stdio-l1-1-0.dll dependency of d:\pyinstaller_bug\venv\scripts\python.exe
4287 WARNING: lib not found: api-ms-win-crt-runtime-l1-1-0.dll dependency of d:\pyinstaller_bug\venv\scripts\python.exe
4503 WARNING: lib not found: api-ms-win-crt-heap-l1-1-0.dll dependency of d:\pyinstaller_bug\venv\scripts\python.exe
4725 WARNING: lib not found: api-ms-win-crt-locale-l1-1-0.dll dependency of d:\pyinstaller_bug\venv\scripts\python.exe
4944 WARNING: lib not found: api-ms-win-crt-math-l1-1-0.dll dependency of d:\pyinstaller_bug\venv\scripts\python.exe
5273 WARNING: lib not found: api-ms-win-crt-conio-l1-1-0.dll dependency of d:\pyinstaller_bug\venv\scripts\python35.dll
5487 WARNING: lib not found: api-ms-win-crt-filesystem-l1-1-0.dll dependency of d:\pyinstaller_bug\venv\scripts\python35.dll
5708 WARNING: lib not found: api-ms-win-crt-convert-l1-1-0.dll dependency of d:\pyinstaller_bug\venv\scripts\python35.dll
5931 WARNING: lib not found: api-ms-win-crt-environment-l1-1-0.dll dependency of d:\pyinstaller_bug\venv\scripts\python35.dll
6150 WARNING: lib not found: api-ms-win-crt-stdio-l1-1-0.dll dependency of d:\pyinstaller_bug\venv\scripts\python35.dll
6375 WARNING: lib not found: api-ms-win-crt-runtime-l1-1-0.dll dependency of d:\pyinstaller_bug\venv\scripts\python35.dll
6600 WARNING: lib not found: api-ms-win-crt-heap-l1-1-0.dll dependency of d:\pyinstaller_bug\venv\scripts\python35.dll
6820 WARNING: lib not found: api-ms-win-crt-locale-l1-1-0.dll dependency of d:\pyinstaller_bug\venv\scripts\python35.dll
7036 WARNING: lib not found: api-ms-win-crt-math-l1-1-0.dll dependency of d:\pyinstaller_bug\venv\scripts\python35.dll
7259 WARNING: lib not found: api-ms-win-crt-process-l1-1-0.dll dependency of d:\pyinstaller_bug\venv\scripts\python35.dll
7474 WARNING: lib not found: api-ms-win-crt-time-l1-1-0.dll dependency of d:\pyinstaller_bug\venv\scripts\python35.dll
7690 WARNING: lib not found: api-ms-win-crt-string-l1-1-0.dll dependency of d:\pyinstaller_bug\venv\scripts\python35.dll
7916 WARNING: lib not found: api-ms-win-crt-convert-l1-1-0.dll dependency of d:\pyinstaller_bug\venv\scripts\VCRUNTIME140.dll
8135 WARNING: lib not found: api-ms-win-crt-stdio-l1-1-0.dll dependency of d:\pyinstaller_bug\venv\scripts\VCRUNTIME140.dll
8360 WARNING: lib not found: api-ms-win-crt-runtime-l1-1-0.dll dependency of d:\pyinstaller_bug\venv\scripts\VCRUNTIME140.dll
8578 WARNING: lib not found: api-ms-win-crt-heap-l1-1-0.dll dependency of d:\pyinstaller_bug\venv\scripts\VCRUNTIME140.dll
8799 WARNING: lib not found: api-ms-win-crt-string-l1-1-0.dll dependency of d:\pyinstaller_bug\venv\scripts\VCRUNTIME140.dll
8801 INFO: Caching module hooks...
8807 INFO: Analyzing main.py
8819 INFO: Loading module hooks...
8819 INFO: Loading module hook "hook-xml.py"...
9081 INFO: Loading module hook "hook-encodings.py"...
9309 INFO: Loading module hook "hook-pydoc.py"...
9322 INFO: Looking for ctypes DLLs
9333 INFO: Analyzing run-time hooks ...
9339 INFO: Looking for dynamic libraries
9565 WARNING: lib not found: api-ms-win-crt-conio-l1-1-0.dll dependency of D:\pyinstaller_bug\venv\Scripts\_ssl.pyd
9792 WARNING: lib not found: api-ms-win-crt-filesystem-l1-1-0.dll dependency of D:\pyinstaller_bug\venv\Scripts\_ssl.pyd
10018 WARNING: lib not found: api-ms-win-crt-convert-l1-1-0.dll dependency of D:\pyinstaller_bug\venv\Scripts\_ssl.pyd
10262 WARNING: lib not found: api-ms-win-crt-environment-l1-1-0.dll dependency of D:\pyinstaller_bug\venv\Scripts\_ssl.pyd
10510 WARNING: lib not found: api-ms-win-crt-stdio-l1-1-0.dll dependency of D:\pyinstaller_bug\venv\Scripts\_ssl.pyd
10740 WARNING: lib not found: api-ms-win-crt-string-l1-1-0.dll dependency of D:\pyinstaller_bug\venv\Scripts\_ssl.pyd
10967 WARNING: lib not found: api-ms-win-crt-utility-l1-1-0.dll dependency of D:\pyinstaller_bug\venv\Scripts\_ssl.pyd
11189 WARNING: lib not found: api-ms-win-crt-runtime-l1-1-0.dll dependency of D:\pyinstaller_bug\venv\Scripts\_ssl.pyd
11422 WARNING: lib not found: api-ms-win-crt-heap-l1-1-0.dll dependency of D:\pyinstaller_bug\venv\Scripts\_ssl.pyd
11689 WARNING: lib not found: api-ms-win-crt-time-l1-1-0.dll dependency of D:\pyinstaller_bug\venv\Scripts\_ssl.pyd
11910 WARNING: lib not found: api-ms-win-crt-stdio-l1-1-0.dll dependency of D:\pyinstaller_bug\venv\Scripts\unicodedata.pyd
12137 WARNING: lib not found: api-ms-win-crt-runtime-l1-1-0.dll dependency of D:\pyinstaller_bug\venv\Scripts\unicodedata.pyd
12356 WARNING: lib not found: api-ms-win-crt-string-l1-1-0.dll dependency of D:\pyinstaller_bug\venv\Scripts\unicodedata.pyd
12741 WARNING: lib not found: api-ms-win-crt-stdio-l1-1-0.dll dependency of D:\pyinstaller_bug\venv\Scripts\_ctypes.pyd
13020 WARNING: lib not found: api-ms-win-crt-string-l1-1-0.dll dependency of D:\pyinstaller_bug\venv\Scripts\_ctypes.pyd
13250 WARNING: lib not found: api-ms-win-crt-runtime-l1-1-0.dll dependency of D:\pyinstaller_bug\venv\Scripts\_ctypes.pyd
13474 WARNING: lib not found: api-ms-win-crt-stdio-l1-1-0.dll dependency of D:\pyinstaller_bug\venv\Scripts\pyexpat.pyd
13696 WARNING: lib not found: api-ms-win-crt-utility-l1-1-0.dll dependency of D:\pyinstaller_bug\venv\Scripts\pyexpat.pyd
13915 WARNING: lib not found: api-ms-win-crt-runtime-l1-1-0.dll dependency of D:\pyinstaller_bug\venv\Scripts\pyexpat.pyd
14139 WARNING: lib not found: api-ms-win-crt-heap-l1-1-0.dll dependency of D:\pyinstaller_bug\venv\Scripts\pyexpat.pyd
14360 WARNING: lib not found: api-ms-win-crt-time-l1-1-0.dll dependency of D:\pyinstaller_bug\venv\Scripts\pyexpat.pyd
14591 WARNING: lib not found: api-ms-win-crt-conio-l1-1-0.dll dependency of D:\pyinstaller_bug\venv\Scripts\_hashlib.pyd
14813 WARNING: lib not found: api-ms-win-crt-convert-l1-1-0.dll dependency of D:\pyinstaller_bug\venv\Scripts\_hashlib.pyd
15032 WARNING: lib not found: api-ms-win-crt-environment-l1-1-0.dll dependency of D:\pyinstaller_bug\venv\Scripts\_hashlib.pyd
15263 WARNING: lib not found: api-ms-win-crt-stdio-l1-1-0.dll dependency of D:\pyinstaller_bug\venv\Scripts\_hashlib.pyd
15484 WARNING: lib not found: api-ms-win-crt-string-l1-1-0.dll dependency of D:\pyinstaller_bug\venv\Scripts\_hashlib.pyd
15703 WARNING: lib not found: api-ms-win-crt-utility-l1-1-0.dll dependency of D:\pyinstaller_bug\venv\Scripts\_hashlib.pyd
15920 WARNING: lib not found: api-ms-win-crt-runtime-l1-1-0.dll dependency of D:\pyinstaller_bug\venv\Scripts\_hashlib.pyd
16141 WARNING: lib not found: api-ms-win-crt-heap-l1-1-0.dll dependency of D:\pyinstaller_bug\venv\Scripts\_hashlib.pyd
16368 WARNING: lib not found: api-ms-win-crt-time-l1-1-0.dll dependency of D:\pyinstaller_bug\venv\Scripts\_hashlib.pyd
16591 WARNING: lib not found: api-ms-win-crt-stdio-l1-1-0.dll dependency of D:\pyinstaller_bug\venv\Scripts\_bz2.pyd
16811 WARNING: lib not found: api-ms-win-crt-runtime-l1-1-0.dll dependency of D:\pyinstaller_bug\venv\Scripts\_bz2.pyd
17034 WARNING: lib not found: api-ms-win-crt-heap-l1-1-0.dll dependency of D:\pyinstaller_bug\venv\Scripts\_bz2.pyd
17252 WARNING: lib not found: api-ms-win-crt-math-l1-1-0.dll dependency of D:\pyinstaller_bug\venv\Scripts\_bz2.pyd
17469 WARNING: lib not found: api-ms-win-crt-string-l1-1-0.dll dependency of D:\pyinstaller_bug\venv\Scripts\_bz2.pyd
17697 WARNING: lib not found: api-ms-win-crt-heap-l1-1-0.dll dependency of D:\pyinstaller_bug\venv\Scripts\_lzma.pyd
17913 WARNING: lib not found: api-ms-win-crt-runtime-l1-1-0.dll dependency of D:\pyinstaller_bug\venv\Scripts\_lzma.pyd
18137 WARNING: lib not found: api-ms-win-crt-runtime-l1-1-0.dll dependency of D:\pyinstaller_bug\venv\Scripts\_socket.pyd
18362 WARNING: lib not found: api-ms-win-crt-runtime-l1-1-0.dll dependency of D:\pyinstaller_bug\venv\Scripts\select.pyd
18366 INFO: Looking for eggs
18366 INFO: Using Python library d:\pyinstaller_bug\venv\scripts\python35.dll
18366 INFO: Found binding redirects:
[]
Traceback (most recent call last):
  File "D:\pyinstaller_bug\venv\Scripts\pyinstaller-script.py", line 9, in <module>
    load_entry_point('PyInstaller==3.5.dev0+baf94fc1b', 'console_scripts', 'pyinstaller')()
  File "d:\pyinstaller_bug\venv\lib\site-packages\PyInstaller\__main__.py", line 111, in run
    run_build(pyi_config, spec_file, **vars(args))
  File "d:\pyinstaller_bug\venv\lib\site-packages\PyInstaller\__main__.py", line 63, in run_build
    PyInstaller.building.build_main.main(pyi_config, spec_file, **kwargs)
  File "d:\pyinstaller_bug\venv\lib\site-packages\PyInstaller\building\build_main.py", line 844, in main
    build(specfile, kw.get('distpath'), kw.get('workpath'), kw.get('clean_build'))
  File "d:\pyinstaller_bug\venv\lib\site-packages\PyInstaller\building\build_main.py", line 791, in build
    exec(code, spec_namespace)
  File "main.spec", line 25, in <module>
    noarchive=True)
  File "d:\pyinstaller_bug\venv\lib\site-packages\PyInstaller\building\build_main.py", line 243, in __init__
    self.__postinit__()
  File "d:\pyinstaller_bug\venv\lib\site-packages\PyInstaller\building\datastruct.py", line 158, in __postinit__
    self.assemble()
  File "d:\pyinstaller_bug\venv\lib\site-packages\PyInstaller\building\build_main.py", line 597, in assemble
    for name, path, typecode in compile_py_files(new_toc, CONF['workpath']):
  File "d:\pyinstaller_bug\venv\lib\site-packages\PyInstaller\utils\misc.py", line 150, in compile_py_files
    with open(obj_fnm, 'rb') as fh:
FileNotFoundError: [Errno 2] No such file or directory: '-'
```

**Comment by htgoebel** [comment]
Thanks for this bug-report. This look like you hit a interesting case :-)
May I ask you to put the code for reproducing into a git-repo. For security reasons I'm not downloding and extracting external data. Thanks.

**Comment by jmakitalo** [comment]
Here it is: https://github.com/jmakitalo/pyinstaller_4034

**Comment by kkumuk** [comment]
I have same problem. How can i resolve it before bug fix?

**Comment by jmakitalo** [comment]
I was able to fix it by adding

```
        # Fix case when fnm represents a namespace.
        if fnm in ('-', None):
            continue
```

on line 128 in file misc.py . I'm not sure if this is correct solution considering everything as I'm not expert in Pyinstaller code.

**Comment by kkumuk** [comment]
Thank you for your help. I could finish the building. There is no problems so far.

**Comment by cdgriffith** [comment]
I also experienced this error when adding the `--debug` option. Thank you @jmakitalo for the workaround!

**Comment by thomasdeneux** [comment]
> I also experienced this error when adding the `--debug` option. Thank you @jmakitalo for the workaround!

same for me

**Comment by danyeaw** [comment]
I also ran in to this issue.
@jmakitalo could you submit a PR with your workaround?

**Comment by MeridianPY** [comment]
Yeah. For me it also was the `--debug` option. Weird, because I remember it working flawlessly before. Seems like it only breaks sometimes...

**Comment by sergiomsilva** [comment]
It seems to me that the problem relies on the `noarchive` option. When I run with `--debug imports --debug bootloader` instead of `--debug all` there is no problem.

**Comment by nimaid** [comment]
> 
> 
> Yeah. For me it also was the `--debug` option. Weird, because I remember it working flawlessly before. Seems like it only breaks sometimes...
> 
> Edit:
> Apparently it works if you build the .exe without the `--debug` option and rebuild it after that with the option added to the command. Works for me...

This does work, just make sure you don't also use the `--clean` option.

**Comment by annesehnal** [comment]
@jmakitalo Thank you so much for the bugfix I would have never figured this out!!

**Comment by sarthak260** [comment]
![image](https://user-images.githubusercontent.com/34007828/89708824-29eee580-d998-11ea-8147-d6b7b7f5233d.png)

i am still facing this issue after adding the --debug=all flag to command line pyinstaller 

Can pls anyone help me out in any way?

**Comment by anthello147** [comment]
I have a problem compiling anything with PyInstaller 4.0 on a Chromium OS virtual machine. I installed PyInstaller using the command `pip install pyinstaller` on a Miniconda installation of Python 3.8. I can't compile even a simple `print("Hello, World!"` because PyInstaller gives me the following error (I have tried installing PyInstaller with Conda with `conda install -c anaconda pyinstaller` but the Conda version still yields the same results):
```
107 INFO: PyInstaller: 4.0
107 INFO: Python: 3.8.3 (conda)
174 INFO: Platform: Linux-4.0.5+-x86_64-Intel-R-_Core-TM-_i5-7200U_CPU_@_2.50GHz-with-glibc2.10
177 INFO: wrote /home/chronos/user/Downloads/test.spec
178 INFO: UPX is not available.
179 INFO: Extending PYTHONPATH with paths
['/home/chronos/user/Downloads', '/home/chronos/user/Downloads']
179 INFO: checking Analysis
179 INFO: Building Analysis because Analysis-00.toc is non existent
179 INFO: Initializing module dependency graph...
184 INFO: Caching module graph hooks...
191 INFO: Analyzing base_library.zip ...
2299 INFO: Processing pre-find module path hook   distutils
2303 INFO: distutils: retargeting to non-venv dir '/usr/local/conda3/lib/python3.8'
4440 INFO: Caching module dependency graph...
4539 INFO: running Analysis Analysis-00.toc
4552 INFO: Analyzing /home/chronos/user/Downloads/test.py
4554 INFO: Processing module hooks...
4554 INFO: Loading module hook "hook-pydoc.py"...
4555 INFO: Loading module hook "hook-xml.py"...
4607 INFO: Loading module hook "hook-xml.etree.cElementTree.py"...
4614 INFO: Loading module hook "hook-distutils.py"...
4619 INFO: Loading module hook "hook-_tkinter.py"...
4699 INFO: checking Tree
4699 INFO: Building Tree because Tree-00.toc is non existent
4699 INFO: Building Tree Tree-00.toc
4711 INFO: checking Tree
4712 INFO: Building Tree because Tree-01.toc is non existent
4712 INFO: Building Tree Tree-01.toc
4719 INFO: Loading module hook "hook-sysconfig.py"...
4729 INFO: Loading module hook "hook-encodings.py"...
4781 INFO: Loading module hook "hook-lib2to3.py"...
4805 INFO: Looking for ctypes DLLs
4840 INFO: Analyzing run-time hooks ...
4843 INFO: Including run-time hook 'pyi_rth_multiprocessing.py'
4850 INFO: Including run-time hook 'pyi_rth__tkinter.py'
4858 INFO: Looking for dynamic libraries
5200 INFO: Looking for eggs
5200 INFO: Python library not in binary dependencies. Doing additional searching...
5209 INFO: Using Python library /usr/local/conda3/lib/libpython3.8.so.1.0
5212 INFO: Warnings written to /home/chronos/user/Downloads/build/test/warn-test.txt
5253 INFO: Graph cross-reference written to /home/chronos/user/Downloads/build/test/xref-test.html
5282 INFO: checking PYZ
5282 INFO: Building PYZ because PYZ-00.toc is non existent
5282 INFO: Building PYZ (ZlibArchive) /home/chronos/user/Downloads/build/test/PYZ-00.pyz
5691 INFO: Building PYZ (ZlibArchive) /home/chronos/user/Downloads/build/test/PYZ-00.pyz completed successfully.
5697 INFO: checking PKG
5701 INFO: Building PKG because PKG-00.toc is non existent
5701 INFO: Building PKG (CArchive) PKG-00.pkg
5714 INFO: Building PKG (CArchive) PKG-00.pkg completed successfully.
5715 INFO: Bootloader /usr/local/conda3/lib/python3.8/site-packages/PyInstaller/bootloader/Linux-64bit/run
5715 INFO: checking EXE
5715 INFO: Building EXE because EXE-00.toc is non existent
5715 INFO: Building EXE from EXE-00.toc
5717 INFO: Appending archive to ELF section in EXE /home/chronos/user/Downloads/build/test/test
Traceback (most recent call last):
  File "/usr/local/conda3/bin/pyinstaller", line 11, in <module>
    sys.exit(run())
  File "/usr/local/conda3/lib/python3.8/site-packages/PyInstaller/__main__.py", line 114, in run
    run_build(pyi_config, spec_file, **vars(args))
  File "/usr/local/conda3/lib/python3.8/site-packages/PyInstaller/__main__.py", line 65, in run_build
    PyInstaller.building.build_main.main(pyi_config, spec_file, **kwargs)
  File "/usr/local/conda3/lib/python3.8/site-packages/PyInstaller/building/build_main.py", line 734, in main
    build(specfile, kw.get('distpath'), kw.get('workpath'), kw.get('clean_build'))
  File "/usr/local/conda3/lib/python3.8/site-packages/PyInstaller/building/build_main.py", line 681, in build
    exec(code, spec_namespace)
  File "/home/chronos/user/Downloads/test.spec", line 20, in <module>
    exe = EXE(pyz,
  File "/usr/local/conda3/lib/python3.8/site-packages/PyInstaller/building/api.py", line 445, in __init__
    self.__postinit__()
  File "/usr/local/conda3/lib/python3.8/site-packages/PyInstaller/building/datastruct.py", line 160, in __postinit__
    self.assemble()
  File "/usr/local/conda3/lib/python3.8/site-packages/PyInstaller/building/api.py", line 606, in assemble
    retcode, stdout, stderr = exec_command_all(
  File "/usr/local/conda3/lib/python3.8/site-packages/PyInstaller/compat.py", line 569, in exec_command_all
    proc = subprocess.Popen(cmdargs, bufsize=-1,  # Default OS buffer size.
  File "/usr/local/conda3/lib/python3.8/subprocess.py", line 854, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/usr/local/conda3/lib/python3.8/subprocess.py", line 1702, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'objcopy'
```
I have no clue what might be causing the error.

**Comment by rokm** [comment]
Fixed by f7168c866d7f2d583bf6eb00fbab647a513f64c1.
